sparse/boot/flash.sh

```diff
@@ -0,0 +1,521 @@
+#
+
+while [ $# -gt 0 ]; do
+    case $1 in
+        --force)
+            FORCE=1
+            ;;
+        --fastboot)
+            shift
+            FASTBOOT_BIN_NAME="$1"
+            ;;
+        --image-path)
+            shift
+            IMAGE_PATH="$1"
+            ;;
+        --blob-path)
+            shift
+            BLOB_PATH="$1"
+            ;;
+        --extra-opts)
+            shift
+            FASTBOOTEXTRAOPTS="$1"
+            ;;
+        --config)
+            shift
+            FLASH_CONFIG="$1"
+            ;;
+        --help)
+            usage
+            exit 0
+            ;;
+        *)
+            echo "Unknown argument '$1'"
+            exit 1
+            ;;
+    esac
+    shift
+done
+
+if [ ! -f "$FLASH_CONFIG" ]; then
+    echo "Configuration file ($FLASH_CONFIG) for flash script not found."
+    exit 1
+fi
+
+source "$FLASH_CONFIG"
+
+if [ ${#FLASH_OPS[@]} -eq 0 ]; then
+    echo "No flash operations defined in $FLASH_CONFIG"
+    exit 1
+fi
+
+UNAME=$(uname)
+
+# Do not need root for fastboot on Mac OS X
+if [ "$UNAME" != "Darwin" ] && [ "$(id -u)" -ne 0 ]; then
+    exec sudo -E bash -c "FORCE=$FORCE FASTBOOT_BIN_NAME=\"$FASTBOOT_BIN_NAME\" IMAGE_PATH=\"$IMAGE_PATH\" BLOB_PATH=\"$BLOB_PATH\" FASTBOOTEXTRAOPTS=\"$FASTBOOTEXTRAOPTS\" FLASH_CONFIG=\"$FLASH_CONFIG\" $0"
+fi
+
+
+case $UNAME in
+    Linux)
+        echo "Detected Linux"
+        ;;
+    Darwin)
+        IFS='.' read -r major minor patch <<< "$(sw_vers -productVersion)"
+        OS_VERSION=$major-$minor
+        echo "Detected Mac OS X - Version: $OS_VERSION"
+        ;;
+    FreeBSD)
+        FASTBOOT_BIN_PATH="/usr/local/bin/"
+        echo "Detected FreeBSD"
+        ;;
+    *)
+        echo "Failed to detect operating system!"
+        exit 1
+        ;;
+esac
+
+
+if [ -z "$FASTBOOT_BIN_NAME" ]; then
+    if ! check_fastboot "fastboot-$UNAME-$OS_VERSION" ; then
+        if ! check_fastboot "fastboot-$UNAME"; then
+            # In case we didn't provide functional fastboot binary to the system
+            # lets check that one is found from the system.
+            if ! which fastboot &>/dev/null; then
+                echo "No 'fastboot' found in \$PATH. To install, use:"
+                echo ""
+                echo "    Debian/Ubuntu/.deb distros:  apt-get install android-tools-fastboot"
+                echo "    Fedora:  yum install android-tools"
+                echo "    OS X:    brew install android-sdk"
+                echo "    FreeBSD: pkg install android-tools-fastboot"
+                echo ""
+                exit 1
+            else
+                FASTBOOT_BIN_NAME=fastboot
+            fi
+        fi
+    fi
+fi
+
+if [ -z "$FLASHCMD_FLASH_BOOT" ]; then
+    ### Detect fastboot flash commands
+
+    # By default use "flash" for boot partition
+    FLASHCMD_FLASH_BOOT="flash"
+
+    # If flash:raw command exists use that instead
+    if ${FASTBOOT_BIN_PATH}${FASTBOOT_BIN_NAME} help | grep -q flash:raw; then
```

Comment:
Even old fastboot has "flash:raw", but it outputs everything to stderr, so this grep is not 100% accurate:)
old fastboot:
stderr: `flash:raw boot <kernel> [ <ramdisk> [ <second> ] ] create bootimage and`
new fastboot:
stdout: `flash:raw PARTITION KERNEL [RAMDISK [SECOND]]`

Reply:
And that's why during this line the old `fastboot help` outputs all help screen to our device's flashing console log.

Reply:
```diff
-    if ${FASTBOOT_BIN_PATH}${FASTBOOT_BIN_NAME} help | grep -q flash:raw; then
+    if ${FASTBOOT_BIN_PATH}${FASTBOOT_BIN_NAME} help 2>&1 | grep -q "flash:raw PARTITION"; then
```